File: pc02/pc02_05/cpp/pc02_05.cpp
/**
 * Shuffle
 * pc110205
 * uva10026
 */

#include <iostream>
#include <string>
#include <vector>
#include <sstream>
using namespace std;

#define NSUITS 4
#define NCARDS 13

enum Suit {
	CLUBS,
	DIAMONDS,
	HEARTS,
	SPADES
};

enum Value {
	TWO,
	THREE,
	FOUR,
	FIVE,
	SIX,
	SEVEN,
	EIGHT,
	NINE,
	TEN,
	JACK,
	QUEEN,
	KING,
	ACE
};

string suitName[] = {"Clubs", "Diamonds", "Hearts", "Spades"};
string valueName[] = {
	"2", "3", "4", "5", "6", "7", "8", "9", "10",
	"Jack", "Queen", "King", "Ace"
};


struct Card {
	Suit suit;
	Value value;
};


int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	vector<Card> oldDeck(NSUITS * NCARDS);
	vector<Card> newDeck(NSUITS * NCARDS);
	vector<vector<int>> shuffles(100, vector<int>(52));

	for (int i = 0; i < NSUITS * NCARDS; i++) {
		oldDeck[i].suit = static_cast<Suit>(i / NCARDS);
		oldDeck[i].value = static_cast<Value>(i % NCARDS);
	}

	newDeck = oldDeck;

	int T;

	cin >> T;

	bool isFirst = true;

	while (T--) {
		if (!isFirst) {
			cout << '\n';
		}

		isFirst = false;

		int n;
		cin >> n;

		for (int k = 0; k < n; k++) {
			for (int i = 0; i < NSUITS * NCARDS; i++) {
				cin >> shuffles[k][i];
			}
		}

		cin.ignore();

		string line;
		while (getline(cin, line) && !line.empty()) {
			stringstream ss(line);
			int shuffleIdx;

			while (ss >> shuffleIdx) {
				for (int i = 0; i < NSUITS * NCARDS; i++) {
					newDeck[i] = oldDeck[shuffles[shuffleIdx - 1][i] - 1];
				}

				oldDeck.swap(newDeck);
			}
		}

		for (int i = 0; i < NSUITS * NCARDS; i++) {
			cout << valueName[oldDeck[i].value] << " of "
				<< suitName[oldDeck[i].suit] << '\n';
		}
	}

	return 0;
}

File: pc02/pc02_05/python/pc02_05.py
"""
Shuffle
pc110205
uva10026
"""

import sys


def main():
    """Main function to process card shuffles."""
    lines = sys.stdin.readlines()

    T = int(lines[0].strip())
    line_idx = 1

    suit_names = ["Clubs", "Diamonds", "Hearts", "Spades"]
    value_names = ["2", "3", "4", "5", "6", "7", "8", "9", "10",
                   "Jack", "Queen", "King", "Ace"]

    first = True

    for _ in range(T):
        if not first:
            print()
        first = False

        n = int(lines[line_idx].strip())
        line_idx += 1

        shuffles = []
        for _ in range(n):
            shuffle = list(map(int, lines[line_idx].strip().split()))
            shuffles.append(shuffle)
            line_idx += 1

        old_deck = [(i // 13, i % 13) for i in range(52)]
        new_deck = old_deck[:]

        while line_idx < len(lines):
            line = lines[line_idx].strip()
            line_idx += 1

            if not line:
                break

            shuffle_indices = list(map(int, line.split()))
            for shuffle_idx in shuffle_indices:
                new_deck = []
                for i in range(52):
                    pos = shuffles[shuffle_idx - 1][i] - 1
                    new_deck.append(old_deck[pos])

                old_deck = new_deck

        for suit, value in old_deck:
            print(f"{value_names[value]} of {suit_names[suit]}")


if __name__ == '__main__':
    main()
